File: Army(c++)/State/WolfState.cpp
#include "WolfState.hpp"
#include "WereWolfState.hpp"
#include "../Basic/Unit.hpp"

WolfState::WolfState(Unit* thisUnit) : thisUnit(thisUnit) {}

void WolfState::changeUnit(Unit* enemy) {
    std::cout << "No ability!!!"  << std::endl;
}

void WolfState::transformTo() {
    this->thisUnit->getHealth().ensureIsAlive();
    
    this->thisUnit->transitionTo(new WereWolfState(this->thisUnit));
    
    this->thisUnit->setWarriorID(WarriorID::WEREWOLF);
    this->thisUnit->getHealth().setDamage(this->thisUnit->getHealth().getDamage()/1.5);
    this->thisUnit->getHealth().setHitPoints(this->thisUnit->getHealth().getHitPoints()/2);
    this->thisUnit->getHealth().setHitPointsLimit(this->thisUnit->getHealth().getHitPointsLimit()/2);
}

File: Army(java)/State/WolfState.java
package State;
import Basic.*;
import EnumID.*;

public class WolfState implements State {
    protected Unit thisUnit;

    public WolfState(Unit unit) {
        this.thisUnit = unit;
    }

    public void printState() {
        System.out.println("\tI'm Wolf!!!");
    }

    public void changeUnit(Unit enemy) throws UnitIsDeadException {
        System.out.println("No ability!!!");
    }

    public void transformTo() throws UnitIsDeadException {
        thisUnit.getHealth().ensureIsAlive();
        
        thisUnit.setState(new WereWolfState(thisUnit));
        
        thisUnit.setWarriorID(WarriorID.WEREWOLF);
        thisUnit.getHealth().setDamage(thisUnit.getHealth().getDamage()/2);
        thisUnit.getHealth().setHitPoints(thisUnit.getHealth().getHitPoints()/2);
        thisUnit.getHealth().setHitPointsLimit(thisUnit.getHealth().getHitPointsLimit()/2);
    }
}
